[SYCL]Add device libraries for C math and complex

jinge90 opened the pull request:
Add device libraries for C math and complex functions based on SYCL compiler's device library infrastructure. Fallback implementation for math and complex functions are also provided. The fallback implementation for C math functions are based on _spirv_ocl* math built-ins and the fallback implementation for C99 complex functions are based on algorithm used in libc++ std::complex. All functions supported are listed in sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst.

asavonic commented on sycl/doc/extensions/C-CXX-StandardLibrary/DeviceLibExtensions.rst:
```diff
@@ -32,3 +32,159 @@ Arguments:
 .. code:
    foo.cpp:42: void foo(int): global id: [0,0,0], local id: [0,0,0] Assertion `buf[wiID] == 0 && "Invalid value"` failed.
+
+cl_intel_devicelib_math
```
Not all functions from math.h are supported, specifically:
  - abs
  - ceil
  - copysign
  - fabs
  - floor
  - fmax
  - fmin
  - llrint
  - llround
  - lrint
  - lround
  - nan
  - nearbyint
  - nexttoward
  - rint
  - round
  - scalbln
  - scalbn
  - trunc

jinge90 replied:
Hi, Andrew. In the long term, we can support them. So, I think we can mention that support for those functions is  in future plan.

asavonic replied:
Please list them in DeviceLibExtensions.rst as well.
These extensions are supposed to be a "contract" between the FE and a device compiler. We should not add more functions to a particular extension, otherwise we invalidate device compilers that support an older version.

jinge90 replied:
Hi, Andrew. So I need to claim that:
1. we don't support all functions in math.h in cl_intel_devicelib_math extension
2. List the functions which we plan to support in the future
in DeviceLibExtensions.rst?
Can I add a "Note:" section in cl_intel_devicelib_math to claim those 2 points? Like following:
Arguments:
Those __devicelib_* functions have the same argument type and return type as corresponding
math functions from <math.h> or <cmath>.

Note:
Currently, not all functions from <math.h>; or <cmath> are supported in this extension. Following functions will be supported in the future:
.. code:
   int    __devicelib_abs(int x);
   float  __devicelib_ceilf(float x);
   double __devicelib_ceil(double x);
   float  __devicelib_copysign(float x, float y);
   double __devicelib_copysign(double x, double y);
   float  __devicelib_fabsf(float x);
   double __devicelib_fabs(double x);
   float  __devicelib_floorf(float x);
   double __devicelib_floor(double x);
   float  __devicelib_fmaxf(float x);
   double __devicelib_fmax(double x);
   float  __devicelib_fminf(float x);
   double __devicelib_fmin(double x);
   float  __devicelib_nextafterf(float x, float y);
   double __devicelib_nextafter(double x, double y);
   float  __devicelib_rintf(float x);
   double __devicelib_rint(double x);
   float  __devicelib_roundf(float x);
   double __devicelib_round(double x);
   float  __devicelib_truncf(float x);
   double __devicelib_trunc(double x);

Thank you very much.

asavonic replied:
> Note:
> Currently, not all functions from <math.h>; or are supported in this extension. Following functions will be supported in the future:
> .. code:
> int __devicelib_abs(int x);
> float __devicelib_ceilf(float x);
> double __devicelib_ceil(double x);
> [...]

Sounds good.

jinge90 replied:
Just added.
Thank you very much.

asavonic commented on sycl/doc/extensions/C-CXX-StandardLibrary/DeviceLibExtensions.rst:
```diff
@@ -32,3 +32,159 @@ Arguments:
    foo.cpp:42: void foo(int): global id: [0,0,0], local id: [0,0,0] Assertion `buf[wiID] == 0 && "Invalid value"` failed.
+
+cl_intel_devicelib_math
+==========================
```
Do you plan to support them as part of this extension later? If so,
then you probably should mention it here.

asavonic commented on sycl/doc/extensions/C-CXX-StandardLibrary/DeviceLibExtensions.rst:
```diff
@@ -32,3 +32,159 @@ Arguments:
+
+Arguments:
+Those __devicelib_* functions have the same arugment type and return type as corresponding
+complex math functions from <complex.h>. The "float __complex__" or "double __complex__" type
```
arugment ->  argument

jinge90 replied:
Fixed.

asavonic commented on sycl/source/detail/devicelib/cmath_wrapper.cpp:
```diff
@@ -0,0 +1,406 @@
+#include "device_math.h"
+extern "C" {
+SYCL_EXTERNAL
+float __attribute__((weak)) scalbnf(float x, int n) {
```
Please mention why we have __attribute__((weak)) here.

jinge90 replied:
OK, I will add comments here and add one execution test to show that the functions provided by users will have the highest priority and will override our "fallback" implementation.

asavonic commented on sycl/source/detail/program_manager/program_manager.cpp:
```diff
@@ -570,7 +578,9 @@ static std::vector<RT::PiProgram> getDeviceLibPrograms(
   std::pair<DeviceLibExt, bool> RequiredDeviceLibExt[] = {
-      {cl_intel_devicelib_assert, /* is fallback loaded? */ false}
+      {cl_intel_devicelib_assert, /* is fallback loaded? */ false},
+      {cl_intel_devicelib_math, /* is fallback loaded? */ false},
```
I think you can omit this "is fallback loaded?" comment.

jinge90 replied:
Fixed.

bader commented on sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst:
```diff
@@ -58,6 +133,31 @@ Example of usage
+         auto res_access1 = buffer1.get_access<sycl_write>(cgh);
+         auto res_access2 = buffer2.get_access<sycl_write>(cgh);
+         cgh.single_task<class DeviceSin>([=]() {
+           res_access1[0] = sinf(0);
```
```suggestion
           res_access1[0] = sinf(0.f);
```

jinge90 replied:
Fixed

bader commented on sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst:
```diff
@@ -58,6 +133,31 @@ Example of usage
+         auto res_access2 = buffer2.get_access<sycl_write>(cgh);
+         cgh.single_task<class DeviceSin>([=]() {
+           res_access1[0] = sinf(0);
+           res_access2[0] = sin(0);
```
```suggestion
           res_access2[0] = sin(0.0);
```

jinge90 replied:
Fixed

bader commented on sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst:
```diff
@@ -58,6 +133,31 @@ Example of usage
+         });
+       });
+     }
+     assert((result_f == 0) && (result_d == 0));
```
```suggestion
     assert((result_f == 0.f) && (result_d == 0.0));
```

jinge90 replied:
Fixed

bader commented on sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst:
```diff
@@ -58,6 +133,31 @@ Example of usage
+   void device_sin_test() {
+     cl::sycl::queue deviceQueue;
+     cl::sycl::range<1> numOfItems{1};
+     float  result_f = -1;
```
```suggestion
     float  result_f = -1.f;
```

jinge90 replied:
Fixed

bader commented on sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst:
```diff
@@ -58,6 +133,31 @@ Example of usage
+     cl::sycl::queue deviceQueue;
+     cl::sycl::range<1> numOfItems{1};
+     float  result_f = -1;
+     double result_d = -1;
```
```suggestion
     double result_d = -1.0;
```

bader commented on sycl/doc/extensions/C-CXX-StandardLibrary/DeviceLibExtensions.rst:
```diff
@@ -32,3 +32,184 @@ Arguments:
+
+Note:
+Currently, not all functions from <math.h> or <cmath> are supported in this extension. Following
+functions will be supported in the future:
```
```suggestion
functions are not supported:
```

bader commented on sycl/doc/extensions/C-CXX-StandardLibrary/DeviceLibExtensions.rst:
```diff
@@ -32,3 +32,184 @@ Arguments:
+
+Semantic:
+Those __devicelib_* functions have the same semantic as corresponding complex math functions
+from <complex.h>. Please refer to: https://en.cppreference.com/w/c/numeric/complex for details.
```
I think it's better to refer to the C++/C standard rather than cppreference web-site.

jinge90 replied:
Yes, I will update them all.

jinge90 replied:
Hi, Alexey.
I changed to "Please refer to ISO/IEC 14882:2011" as C++ standard document is not free, we can provide a link to online doc here.

bader commented on sycl/include/CL/sycl/detail/program_manager/program_manager.hpp:
```diff
@@ -42,7 +42,9 @@ using DeviceImage = pi_device_binary_struct;
-  cl_intel_devicelib_assert = 0
+  cl_intel_devicelib_assert = 0,
+  cl_intel_devicelib_math,
+  cl_intel_devicelib_complex
```
What is the logic behind separating device library?
Is there any document describing where should new device function go?

jinge90 replied:
It seems that we don't have such documentation.  In current device lib design, if device provides native functions to override DPC++ compiler's fallback implementation, our fallback device lib won't be loaded by sycl runtime. Suppose we put everything in just one device lib including complex, cmath...., if device only provide native support for cmath functions but nothing for complex and assert, the fallback devicelib won't be loaded and the complex/assert will not work.

bader replied:
Please, document the design.

jinge90 replied:
Already documented in C-CXX-StandardLibrary.rst.

bader commented on sycl/doc/extensions/C-CXX-StandardLibrary/DeviceLibExtensions.rst:
```diff
@@ -32,3 +32,184 @@ Arguments:
+Semantic:
+Those __devicelib_* functions have the same semantic as corresponding math functions
+from <math.h> or <cmath>. Please refer to: http://www.cplusplus.com/reference/cmath/
+for details.
```
> The fallback implementation for C math functions are based on `__spirv_ocl*` math built-ins.

AFAIK, semantics of `__spirv_ocl*` math built-ins deviates from C math library semantics (see section 7 of OpenCL specification).
It means that the same function will behave differently depending on whether native or fallback function implementation is used. How user is supposed to reason about function behavior?

jinge90 replied:
This is a good question. In fact, all math functions supported currently are wrappers of corresponding sycl ocl math builtins. And sycl::* math functions are wrappers of corresponding __spirv_ocl_* math builtin too, can we say: Those __devicelib_* functions have the same semantic as corresponding sycl math functions? And we can ask users to refer to sycl spec.
Thank you very much.

bader replied:
I think it should be the opposite. Fallback device implementation of C/C++ math functions must follow C/C++ math functions semantics on the host to be consistent. 
SYCL math functions semantics is different and currently aligned with `__spirv_ocl*` functions.

andykaylor replied:
When you write a kernel, the semantics should be the same regardless of where the kernel is executed. I don't see how we can promise C/C++ math function semantics because we can't, in general, do things like setting errno.

bader replied:
> I don't see how we can promise C/C++ math function semantics because we can't, in general, do things like setting errno.

I think all deviations from C/C++ math function semantics must be documented.
@andykaylor, do you have a full list of things which can't be supported on the accelerators?

jinge90 replied:
Hi, Alexey and Andy. I also think we need to document the deviations from C/C++ and I need to spend some time looking up the Linux manual and SYCL spec to find those deviations. Currently, the first things comes to my mind is some math functions are supposed to set errno and float-point exception(do we support fp-exeception in SYCL kernel?).
And we have "cl::sycl::rint" in SYCL math functions, the corresponding math functions in Linux is rint and following description is from Linux man page:

"The  rint(),  rintf(),  and rintl() functions do the same, but will raise the inexact exception (FE_INEXACT, checkable via fetestexcept(3)) when the result differs in value from the argument."
I looked into SYCL spec and didn't claim cl::sycl::rint would raise exception or not. Does sycl or OpenCL spec has clear defined whether such FP related exception is supported?
Thank you very much.

bader replied:
I think most of the differences are covered in section 7 of OpenCL specification.

jinge90 replied:
To functions in current math device libraries, I searched the libm document and found most of those functions may set errno and raise FP-exceptions and I look up the OCL spec and found FP-exception is disabled in OCL. We need document that "errno and float-point exception are not supported in SYCL device library".

andykaylor replied:
@bader As far as I know the complete list of things that can't be supported is errno and FP exception handling. The standard requires an identifier, math_errhandling, which indicates what form of error handling is supported. If we do not set the MATH_ERRNO bit in this value, the math functions are not expected to correctly set errno. If we do not set MATH_ERREXCEPT, the math functions are not expected to raise FP exceptions. Strictly speaking, the standard requires conforming implementations to set one or both of these values, but I think it we set neither and document this as a known divergence from the standard our behavior should be well understood.

The OpenCL specification defines some requirements for boundary condition results that go beyond the C99 requirements. I assume the built-in implementations must already be meeting these requirements for all devices we support.

jinge90 replied:
Already modified the DeviceLibExtensions.rst according to Andy's suggestion.

bader commented on sycl/test/devicelib/cmath_test.cpp:
```diff
@@ -0,0 +1,85 @@
+// UNSUPPORTED: windows
```
Could you clarify what is the problem with the Windows support?

jinge90 replied:
I found device lib infrastructure have problems supporting multiple fallback device lib.
Not matter how many fallback spv files being loaded and linked with users's device code by sycl runtime, only the first one can work on Windows. I have reported this issue to Andrew.
Another block is MSVC complex implementation depends some MSVC internal math functions, we haven't supported them in device lib.

asavonic replied:
> I found device lib infrastructure have problems supporting multiple fallback device lib.
> Not matter how many fallback spv files being loaded and linked with users's device code by sycl runtime, only the first one can work on Windows. I have reported this issue to Andrew.

I think it was an early revision of the patch, and a few things changed since then. I wonder if this is still a problem.

jinge90 replied:
Hi, Andrew. I tried the latest intel/llvm on Window and the problem still exists.

bader replied:
Please, add a TODO comment with clarification of the problem.

jinge90 replied:
OK. I will also file 2 issues for: 1. supporting multiple devicelib on Windows for sycl runtime devicelib infrastructure; 2. provide MSVC internal functions required by complex in device lib;
I am working on MSVC internal functions now.

jinge90 replied:
Added "TODO" for Windows support.

bader commented on sycl/test/devicelib/cmath_test.cpp:
```diff
@@ -0,0 +1,85 @@
+// UNSUPPORTED: windows
+// RUN: %clangxx -fsycl -c %s -o %t.o
+// RUN: %clangxx -fsycl %t.o %llvm_build_libs_dir/libsycl-cmath.o -o %t.out
```
As this is standard library functionality, it should be supported by the compiler, the same way it's supported for regular C/C++. Do you agree?
If so, device side linking must be done implicitly.
+@mdtoguchi.

jinge90 replied:
Yes, I have discussed with Andrew. We both think this should be done implicitly just link libc/libm.a in normal CPU program.

bader replied:
Are you going to fix that?
I don't fully understand how many device libraries you are planning to have, but ideally we should avoid changing driver each time new device library is added.

jinge90 replied:
Hi, Alexey. For comlex/cmath, we don't need any other device libraries. However, if any other developers want to enable other functionalities they may want to add other device libraries.

bader replied:
As I mentioned here: https://github.com/intel/llvm/pull/975#discussion_r364734220, we should document our design decision. 
It's not clear to me, why we keep double precision math functions as part of math library. 
NOTE: not all accelerators has native support for double precision arithmetic.
Does it make sense to move double precision functions to a separate library?

bader replied:
As to original question with driver changes - they can be done in a separate patch.
Please, make sure this request is not lost if you decide to fix it later.

jinge90 replied:
Hi, Alexey.
To the driver changes, I think both libsycl-glibc.o and libsycl-complex.o/libsycl-math.o needs to be added and we need to make request to driver owner.
To double precision, I didn't get your point very well. Do you mean only float type math and complex functions are needed in fallback device libraries as some accelerators may not support double? If so, can we move all double type math/complex functions' implementation to libsycl-cmath.o and libsycl-complex.o? Then, those double type implementations will be linked with users' device code in compiling time and will exist in final device code only users really depend on them.
Thank you.

bader replied:
> Do you mean only float type math and complex functions are needed in fallback device libraries as some accelerators may not support double?

No. I'm asking about the reasoning of splitting device library into different parts. It's not clear to me. I might imagine that it might be split differently. I can suggest anything only after I understand the logic behind the current proposal.

jinge90 replied:
Do you ask about why split the device libraries based on functionality(one for cmath and another for complex math) but not based on other criterion?

bader replied:
Yes.
Also where I can find definition of "split ... based on functionality"? C/C++ specification?

jinge90 replied:
Hi, Alexey.
Before doing the cmath/complex support, I consulted with Andy and Andrew about the granularity of the device libraries and got the feedback that we can split all those functionalities into 2 device libraries: libsycl-cmath and libsycl-complex. And in fact, I think that the granularity for device libraries is a kind of trade off. In fact, for the most flexibility, we should provide one device library for each function provided, so backend compiler can choose any subset of all those functions to override. However, it is not practical to do so as we will have too many device libraries. Putting everything in just one device library is not a good idea too, as I mentioned before, backend compiler will lose other fallback device functions if only wants to support a subset of all functions. So, we have to do some trade off and I think "split...based on functionality" is not defined by any spec but a kind of experience. In CPU platform, we have many libraries such as libc, libm, libstdc++...., libc includes many functions used for C runtime, all math related functions are in libm and libstdc++ will include some functions  for std, we can use the similar way for device library. 
@andykaylor and @asavonic , do you have any comments?
Thank you very much.

asavonic replied:
Right, the "split" is arbitrary, but we try to split how it makes sense. A separate library for each function does not make sense. A single library for *all* functions defeats the purpose of having extensions. C and C++ standard library groups functions and classes by purpose (math.h and compex.h), so we use this as a baseline.

bader replied:
> C and C++ standard library groups functions and classes by purpose (math.h and compex.h), so we use this as a baseline.

I think something like this must be put into our architecture documentation, but with reference to the spec, rather than particular implementation. The spec defines headers, which slit stadard library declarations.

asavonic replied:
> > C and C++ standard library groups functions and classes by purpose (math.h and compex.h), so we use this as a baseline.
> 
> I think something like this must be put into our architecture documentation, but with reference to the spec, rather than particular implementation. The spec defines headers, which slit stadard library declarations.

No strict rule is intended here. It is not clear why you want to define a formal criteria on how the split should be made.

bader replied:
> > > C and C++ standard library groups functions and classes by purpose (math.h and compex.h), so we use this as a baseline.
> > 
> > 
> > I think something like this must be put into our architecture documentation, but with reference to the spec, rather than particular implementation. The spec defines headers, which slit stadard library declarations.
> 
> No strict rule is intended here. It is not clear why you want to define a formal criteria on how the split should be made.

I do not say it's "formal criteria", I want us to document this design decision to keep the implementation sane and consistent. This decision is not obvious from current implementation and if someone else will add new functionality the design must be clear.

andykaylor replied:
I think we need a device library, implemented as wrappers, for all of the functions that correspond to OpenCL built-ins. The fact that the devices may have native support for these is irrelevant. We need the library to perform the mapping from libm call to OpenCL built-in. The alternative is to perform this mapping in the compiler, but we've discussed that and decided the device library is preferable.

I think it would make sense to have a division of the device library for non-complex functions between those with OpenCL equivalents and those without. My suggestion for a separate complex library was based on the idea that some devices might make the decision that complex support was not a priority and the fallback implementations were acceptable.

jinge90 replied:
Already added the design decision in C-CXX-StandardLibrary.rst.

jinge90 commented on sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst:
```diff
@@ -23,10 +23,85 @@ or, in case of Windows:
+supported. The device libraries for <math.h> and <complex.h> are
+ready for Linux and Windows support will be added in the future.
+Not all functions from <math.h> are supported right now, following
+math functions will be supported in the future:
```
Fixed

bader commented on sycl/doc/extensions/C-CXX-StandardLibrary/DeviceLibExtensions.rst:
```diff
@@ -32,3 +32,158 @@ Arguments:
+
+Semantic:
+Those __devicelib_* functions have the same semantic as corresponding math functions
+from <math.h> or <cmath>. Please refer to: http://www.cplusplus.com/reference/cmath/
```
Please, replace all references to web-sites citing C++ spec with direct references to the spec.

jinge90 commented:
Hi, @bader  @asavonic , @andykaylor 
I made a little investigation today and find it is not good to add any device library into clang driver's linking process as default. All device libraries(the wrapper libraries such as libsycl-glibc/cmath/complex) contain LLVM IR code for our wrapper functions and will be linked with users' device code by llvm-link before the llvm-spirv translation if we add them as default. And llvm-link will put everything into final LLVM bc file and then into final SPIR-V used for jit during runtime no matter whether those functions are used by users or not. Although those wrapper libraries don't contain as much code as fallback device library, they may still have big impact to some very small SYCL kernels because we have larger SPIR-V image and the jit overhead will increase. What's more important, I think not many users have strong requirement for those device libraries.
If someone really want to use any device libraries, they can manually link the library with their code.
But we need to document how to use the device library in details.
Thank you very much.

jinge90 commented:
> @jinge90, make sure this patch doesn't introduce any regressions.
Hi, @asavonic 
This regression is due to the overhead of sycl runtime loading fallback spirv and current device library infrastructure can't fix this as we will always load all fallback spv files no matter whether they are really needed or not.
I will try to parse user's device code to see if any dependency on device library exists or not and we can skip all device library if they are not needed by user's code.
As only one case on CPU platform are impacted, can we submit this patch first?

jinge90 commented:
Hi, all. I updated the patch and split device libraries, fallback spir-v libraries and devicelib extension into 2 parts for float and double support. sycl runtime will load fallback spir-v libraries for double support only if underlying device has "cl_khr_fp64" extension, this should fix the problem on some platform without fp64 support.
Hi, @asavonic , could you help me review whether the changes in program_manager.cpp are correct or not? Thank you very much.

bader commented:
@jinge90, could you resolve, merge conflict, please?

jinge90 commented:
> @jinge90, could you resolve, merge conflict, please?

Hi, @bader 
I have rebased and solved the merge conflicts.
Thank you very much.

s-kanaev commented on sycl/source/detail/devicelib/device_math.h:
```diff
@@ -0,0 +1,356 @@
+#ifndef __SYCL_CMATH_WRAPPER_H__
+#define __SYCL_CMATH_WRAPPER_H__
+
+double __spirv_ocl_log(double);
```
All these declarations should reside in spirv_ops.hpp, shouldn't they?

bader commented on sycl/source/detail/devicelib/CMakeLists.txt:
```diff
@@ -45,8 +86,52 @@ add_custom_command(OUTPUT ${binary_dir}/libsycl-fallback-cassert.spv
-add_custom_target(devicelib-obj DEPENDS ${devicelib-obj-file})
-add_custom_target(devicelib-spv DEPENDS ${binary_dir}/libsycl-fallback-cassert.spv)
+add_custom_command(OUTPUT ${binary_dir}/libsycl-fallback-complex.spv
+                   COMMAND ${clang} -S -fsycl-device-only -fno-sycl-use-bitcode
```
I think we should remove `-S` here and from all custom commands below.
If I understand it correctly, today compiler dosn't support it, but it implies textual format of SPIR-V file whereas this command must produce binary format.
